internal/graph/check.go

```diff
@@ -0,0 +1,536 @@
+package graph
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/openfga/openfga/internal/dispatcher"
+	"github.com/openfga/openfga/internal/validation"
+	"github.com/openfga/openfga/pkg/storage"
+	"github.com/openfga/openfga/pkg/tuple"
+	"github.com/openfga/openfga/pkg/typesystem"
+	openfgapb "go.buf.build/openfga/go/openfga/api/openfga/v1"
+)
+
+type setOperatorType int
+
+const (
+	unionSetOperator setOperatorType = iota
+	intersectionSetOperator
+	exclusionSetOperator
+)
+
+type checkOutcome struct {
+	resp *openfgapb.CheckResponse
+	err  error
+}
+
+// ConcurrentChecker implements Check in a highly concurrent and localized manner. The
+// Check resolution is limited per branch of evaluation by the concurrencyLimit.
+type ConcurrentChecker struct {
+	ds               storage.RelationshipTupleReader
+	dispatcher       dispatcher.CheckDispatcher
+	concurrencyLimit uint32
+}
+
+// NewConcurrentChecker constructs a ConcurrentChecker that can be used to evaluate a Check
+// request locally and with a high degree of concurrency.
+func NewConcurrentChecker(
+	ds storage.RelationshipTupleReader,
+	concurrencyLimit uint32,
+) *ConcurrentChecker {
+	checker := &ConcurrentChecker{ds: ds, concurrencyLimit: concurrencyLimit}
+	checker.dispatcher = checker // todo: replace with a different CheckDispatcher once we support remote dispatching
```

Comment:
What is weird about it? Just the variable names used?

Reply:
Hmm no, conceptually... i'm not sure how to explain it... you are creating an object and then assigning a property to it that is the object itself?

Reply:
> you are creating an object and then assigning a property to it that is the object itself?

This also seems weird to me.

Reply:
You cannot initialize an object's field with the object that you are trying to initialize. How else would you recommend this be done?

It may help to understand the vision here...

Eventually the wrapped `dispatcher` will be the delegate that remote dispatching gets routed to. Since we don't yet have remote dispatching, the delegate simply becomes itself (the `LocalChecker`).

Reply:
Can we get rid of the concept of dispatching in this PR? It's not relevant now

Reply:
I don't think removing it in this PR is warranted, because some work has been put into it already. Why would we remove it if it's been done well compositionally? One of the reasons for the refactor _was_ to set us up so that we can dispatch in the near future if we start going that direction architecturally. In the meantime, the dispatching is just a loopback mechanism that dispatches to itself (e.g. no remote dispatching).

This code sets us up to be able to dispatch requests to peers over a remote RPC (or similar) protocol and it adds little to no overhead compositionally in the scope of this PR. This would also be a pretty significant change to this PR, which is already a reasonably sized PR and has already had rounds of review.

Reply:
I think what Jon says above is valid here. We are moving in this direction and have it on our roadmap, so I don't think it hurts to introduce it in this PR. In fact, I think it sets us up nicely because check was written around it.

Reply:
Ok, I tried to meet us all in the middle somewhere. I removed the notion of dispatching but kept things behind some interfaces so that if we ever do add the notion of dispatching it's super easy to port to 👍

Reply:
Resolving this conversation for now. If you have further comments on the new structure, please start a new discussion.